### ISSUE TITLE
Trezor signing fails in multisig wallets in some cases (with fix!)

### ISSUE THREAD

**Issue by Mqrius** [spesmilo/electrum#2550]
tl:dr; It no work because Trezor plugin keystore doesn't sort public keys before signing.

———

Took me a while to figure this out, so bear with me:

I have a 2-of-2 wallet. One of the signers is a Trezor device, the other is a standard wallet from seed.
When I send a transaction via the normal GUI, Electrum first requests and receives the signatures for the inputs from the Trezor, and then generates signatures from the wallet seed.
In the first step of that, Electrum attempts to verify the signatures that came from Trezor. This happens in electrum/transaction.py in function update_signatures.

Now, this works for some signed inputs, and not for others, even in the same transaction. The distinction is: The transaction inputs for which it doesn't work have their public keys not in a sorted order. During the verification process, Electrum sorts the pubkeys via get_sorted_pubkeys(txin). If this shuffles the pubkeys, Electrum is unable to verify Trezor's signature. Sometimes this even causes an undocumented error when during the attempted verification an invalid EC Point is generated.

If I comment out the line in get_sorted_pubkeys that does the actual sorting
`pubkeys, x_pubkeys = zip(*sorted(zip(pubkeys, x_pubkeys)))`
then Electrum is suddenly able to verify Trezor's signatures — but in the end an invalid transaction is generated apparently despite correct signatures. Then again, I don't know what I have and haven't messed up if I remove the ability to sort pubkeys.

The problem lies with the Trezor plugin code. Electrum's keystore objects sort transactions, but the Trezor keystore doesn't. This is easily fixed, as the transaction object has a method that returns sorted public keys — with an (undocumented?) side effect of destructively updating the transaction object's keys to the sorted version.

```
--- electrum-plugins/trezor/plugin.py	2017-06-23 17:48:16.020186804 +0100
+++ electrum-plugins/trezor/plugin.py_old	2017-06-23 17:47:59.472077345 +0100
@@ -53,6 +53,7 @@
         # path of the xpubs that are involved
         xpub_path = {}
         for txin in tx.inputs():
+            pubkeys, x_pubkeys = tx.get_sorted_pubkeys(txin)
             tx_hash = txin['prevout_hash']
             prev_tx[tx_hash] = txin['prev_tx'] 
             for x_pubkey in txin['x_pubkeys']:
```

Now the reason this might not have been noticed before:
It requires a multi-sig wallet with at least 2 of N, in which the Trezor keystore is called first. (If a different keystore is called first, the get_sorted_pubkeys will destructively update the transaction to have sorted pubkeys)

Anyways. Yay, I can transfer my coins :)

**Comment by bauerj** [comment]
Nice! Can you send that in as a pull request?

**Comment by ecdsa** [comment]
thanks a lot for that
